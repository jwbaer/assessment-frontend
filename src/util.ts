import { Component, List, PageApiResult, PageSpec, Variable, VariableDeclaration } from "./types";

export const convertPageApiResultToPageSpec = (apiResult: PageApiResult): PageSpec => {
    const componentIndex: Map<number, Component> = new Map(
        apiResult.components.map((obj: Component) => [obj.id, obj])
    );
    const listIndex: Map<number, List> = new Map(
        apiResult.lists.map((obj: List) => [obj.id, obj])
    );
    const variableIndex: Map<string, Variable> | undefined = apiResult.variables
        ? new Map(
            apiResult.variables.map((obj: VariableDeclaration) => [
                obj.name,
                { name: obj.name, type: obj.type, value: obj.initialValue },
            ])
        )
        : undefined;

    return({ componentIndex, listIndex, variableIndex });
}
