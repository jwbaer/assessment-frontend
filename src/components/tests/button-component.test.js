import { fireEvent, queryByAltText, queryByTestId, render, screen } from "@testing-library/react";

import ButtonComponent from "../button-component";
import ComponentPanel from "../component-panel";
import ListPanel from "../list-panel";
import PageContext from "../../contexts/page-context";
import { convertPageApiResultToPageSpec } from "../../util";

describe("<ButtonComponent /> spec", () => {
    const apiResult= {
        variables: [
            {
                name: 'show_image',
                type: 'string',
                initialValue: 'hide',
            },
        ],
        lists: [
            {
                id: 0,
                components: [1, 2]
            },
            {
                id: 1,
                components: [3]
            }
        ],
        components: [
            {
                id: 1,
                type: 'button',
                options: {
                    text: 'Show',
                    icon: 'show',
                    variable: 'show_image',
                    value: 'show',
                },
            },
            {
                id: 2,
                type: 'condition',
                options: {
                    variable: 'show_image',
                    value: 'show',
                },
                children: 1,
            },
            {
                id: 3,
                type: "image",
                options: {
                    src: "/locations/new-york.png",
                    alt: "Cartoon of New York skyline",
                },
            },
        ],
    };

    let pageSpec = convertPageApiResultToPageSpec(apiResult);

    const setPageSpec = jest.fn((spec) => pageSpec = spec);

    it("renders", () => {
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec}}>
                <ButtonComponent component={pageSpec.componentIndex.get(1)} />
            </PageContext.Provider>
        );

        expect(screen.getByText(/Show/i)).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders through the component panel", () => {
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec}}>
                <ComponentPanel componentId={1} />
            </PageContext.Provider>
        );

        expect(screen.getByText(/Show/i)).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders all components as list, and changes value when clicked on", () => {
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec}}>
                <ListPanel listId={0} />
            </PageContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();
        expect(screen.getByText(/Show/i)).toBeDefined();
        fireEvent.click(screen.getByTestId("button-component-1"));
        expect(setPageSpec).toHaveBeenCalled();
        expect(pageSpec.variableIndex.get('show_image').value).toBe('show');
    });

});
