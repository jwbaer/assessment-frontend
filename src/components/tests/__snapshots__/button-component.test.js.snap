// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonComponent /> spec renders 1`] = `
<DocumentFragment>
  <div
    class="component button"
    data-testid="button-component-1"
  >
    <div
      class="label"
    >
      Show
    </div>
    <div
      class="icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M255.66 112c-77.94 0-157.89 45.11-220.83 135.33a16 16 0 00-.27 17.77C82.92 340.8 161.8 400 255.66 400c92.84 0 173.34-59.38 221.79-135.25a16.14 16.14 0 000-17.47C428.89 172.28 347.8 112 255.66 112z"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
        />
        <circle
          cx="256"
          cy="256"
          fill="none"
          r="80"
          stroke-miterlimit="10"
          stroke-width="32"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ButtonComponent /> spec renders all components as list, and changes value when clicked on 1`] = `
<DocumentFragment>
  <div
    class="list"
  >
    <div
      class="component button"
      data-testid="button-component-1"
    >
      <div
        class="label"
      >
        Show
      </div>
      <div
        class="icon"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M255.66 112c-77.94 0-157.89 45.11-220.83 135.33a16 16 0 00-.27 17.77C82.92 340.8 161.8 400 255.66 400c92.84 0 173.34-59.38 221.79-135.25a16.14 16.14 0 000-17.47C428.89 172.28 347.8 112 255.66 112z"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
          <circle
            cx="256"
            cy="256"
            fill="none"
            r="80"
            stroke-miterlimit="10"
            stroke-width="32"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ButtonComponent /> spec renders through the component panel 1`] = `
<DocumentFragment>
  <div
    class="component button"
    data-testid="button-component-1"
  >
    <div
      class="label"
    >
      Show
    </div>
    <div
      class="icon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M255.66 112c-77.94 0-157.89 45.11-220.83 135.33a16 16 0 00-.27 17.77C82.92 340.8 161.8 400 255.66 400c92.84 0 173.34-59.38 221.79-135.25a16.14 16.14 0 000-17.47C428.89 172.28 347.8 112 255.66 112z"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
        />
        <circle
          cx="256"
          cy="256"
          fill="none"
          r="80"
          stroke-miterlimit="10"
          stroke-width="32"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
